csharp/src/Ice/Reference.cs

```diff
@@ -1210,35 +1247,59 @@ internal async ValueTask<IRequestHandler> GetRequestHandlerAsync(CancellationTok
             return properties;
         }
 
-        // Marshal the reference.
+        // Marshal the non-null reference.
         internal void Write(OutputStream ostr)
         {
             if (IsFixed)
             {
+                // TODO: should be true only for the 1.1 encoding once we add Fixed support in the 2.0 encoding
                 throw new NotSupportedException("cannot marshal a fixed proxy");
             }
 
-            Identity.IceWrite(ostr);
-            ostr.WriteFacet(Facet);
-            ostr.WriteByte((byte)InvocationMode);
-            ostr.WriteBool(false); // secure option, always false (not used)
-            ostr.WriteByte((byte)Protocol);
-            ostr.WriteByte(0);
-            ostr.WriteByte(Encoding.Major);
-            ostr.WriteByte(Encoding.Minor);
-
-            ostr.WriteSize(Endpoints.Count);
-            if (Endpoints.Count > 0)
+            if (ostr.Encoding == Encoding.V11)
             {
-                Debug.Assert(AdapterId.Length == 0);
-                foreach (Endpoint endpoint in Endpoints)
+                Identity.IceWrite(ostr);
+                var proxyData = new ProxyData11(Facet.Length > 0 ? new string[] { Facet } : Array.Empty<string>(),
+                                                InvocationMode,
+                                                secure: false,
+                                                Protocol,
+                                                protocolMinor: 0,
+                                                Encoding);
+                proxyData.IceWrite(ostr);
+                ostr.WriteSequence(Endpoints, (ostr, endpoint) => ostr.WriteEndpoint(endpoint));
```

Comment:
Previously we only write the endpoint sequence when it was not empty

Reply:
I believe I kept the 1.1 encoding unchanged. If the sequence is empty, ostr.WriteSequence writes "0" (1 byte).

Reply:
You are right I confuse because previously we didn't use WriteSequence, and marshal it manually